runtimes/rust/tests/runtime.rs

```diff
@@ -0,0 +1,1249 @@
+        self.events
+            .push(RuntimeEvent::AfterTransactions(height, instance_id));
+        self.runtime.after_transactions(context, instance_id)
+    }
+
+    fn after_commit(&mut self, snapshot: &dyn Snapshot, mailbox: &mut Mailbox) {
+        let height = CoreSchema::new(snapshot).next_height();
+        self.events.push(RuntimeEvent::AfterCommit(height));
+        self.runtime.after_commit(snapshot, mailbox);
+    }
+
+    fn shutdown(&mut self) {
+        self.events.push(RuntimeEvent::Shutdown);
+        self.runtime.shutdown();
+    }
+}
+
+impl WellKnownRuntime for Inspected<RustRuntime> {
+    const ID: u32 = RustRuntime::ID;
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct DeployArtifact {
+    test_service_artifact: ArtifactId,
+    spec: Vec<u8>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct StartService {
+    spec: InstanceSpec,
+    constructor: Vec<u8>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct StopService {
+    instance_id: InstanceId,
+}
+
+impl_binary_value_for_bincode! { DeployArtifact, StartService, StopService }
+
+#[exonum_interface]
+trait ToySupervisor<Ctx> {
+    type Output;
+
+    fn deploy_artifact(&self, context: Ctx, request: DeployArtifact) -> Self::Output;
+    fn start_service(&self, context: Ctx, request: StartService) -> Self::Output;
+    fn stop_service(&self, context: Ctx, request: StopService) -> Self::Output;
+}
+
+#[derive(Debug, ServiceFactory, ServiceDispatcher)]
+#[service_dispatcher(implements("ToySupervisor"))]
+#[service_factory(artifact_name = "toy_supervisor", artifact_version = "0.1.0")]
+struct ToySupervisorService;
+
+impl ToySupervisor<CallContext<'_>> for ToySupervisorService {
+    type Output = Result<(), ExecutionError>;
+
+    fn deploy_artifact(
+        &self,
+        mut context: CallContext<'_>,
+        request: DeployArtifact,
+    ) -> Self::Output {
+        context
+            .supervisor_extensions()
+            .start_artifact_registration(request.test_service_artifact, request.spec)
+    }
+
+    fn start_service(&self, mut context: CallContext<'_>, request: StartService) -> Self::Output {
+        context
+            .supervisor_extensions()
+            .initiate_adding_service(request.spec, request.constructor)
+    }
+
+    fn stop_service(&self, mut context: CallContext<'_>, request: StopService) -> Self::Output {
+        context
+            .supervisor_extensions()
+            .initiate_stopping_service(request.instance_id)
+    }
+}
+
+impl Service for ToySupervisorService {}
+
+impl DefaultInstance for ToySupervisorService {
+    const INSTANCE_ID: u32 = SUPERVISOR_INSTANCE_ID;
+    const INSTANCE_NAME: &'static str = "supervisor";
+
+    fn default_instance(&self) -> InstanceInitParams {
+        self.artifact_id()
+            .into_default_instance(Self::INSTANCE_ID, Self::INSTANCE_NAME)
+            .with_constructor(Init::default())
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct Init {
```

Comment:
Nit: Since we don't care about future / backward compatibility in tests, it could make sense to use `String` directly.